### PR TITLE
Mir & BetterC rework, stage 1

### PR DESCRIPTION
**Pull request by 9il**
First step, remove `instance` as object member. It is `enum` now. 

Next PR: convert all methods of basic allocatros to templates.

This will require a major release.

### CODE REVIEW

File: source/stdx/allocator/building_blocks/segregator.d [hunk 1]
```diff
@@ -190,12 +190,15 @@ struct Segregator(size_t threshold, SmallAllocator, LargeAllocator)
...
+            static if (!hasMember!(typeof(this), "instance"))
+                return .reallocate(this, b, s);
+            else
+                return .reallocate(instance, b, s);
         }
 
         static if (hasMember!(SmallAllocator, "alignedReallocate")
                 || hasMember!(LargeAllocator, "alignedReallocate"))
-        bool reallocate(ref void[] b, size_t s)
+        bool alignedReallocate(ref void[] b, size_t s)
```
Comment (9il) [inline source/stdx/allocator/building_blocks/segregator.d:201]: Fix a bug with wrong function name

File: source/stdx/allocator/building_blocks/null_allocator.d [hunk 2]
```diff
@@ -20,48 +20,48 @@ struct NullAllocator
...
-    bool deallocateAll() shared { return true; }
+    static bool deallocateAll() { return true; }
     /**
     Returns $(D Ternary.yes).
     */
-    Ternary empty() shared const { return Ternary.yes; }
+    static Ternary empty() { return Ternary.yes; }
     /**
-    Returns the $(D shared) global instance of the $(D NullAllocator).
+    static Returns the $(D static) global instance of the $(D NullAllocator).
```
Comment (wilzbach) [inline source/stdx/allocator/building_blocks/null_allocator.d]: Accidental static here

**Review by wilzbach — APPROVED** [review]
I don't see any problem in this module diverging from Phobos.
We should tag the next release as 3.x then though.

File: source/stdx/allocator/building_blocks/null_allocator.d (hunk 2 above)
Comment (9il) [reply source/stdx/allocator/building_blocks/null_allocator.d]: thanx, fixed

**Comment by ghost** [comment]
if peole are using `~>` correctly 2.78.0 can be the new tag too.

**Comment by wilzbach** [comment]
> if peole are using `~>` correctly 2.78.0 can be the new tag too.

Yeah, but this could lead to confusing that the changes are coming from Phobos (e.g. the 2.078 Phobos release).
I don't think it hurts to release the new changes as 3.0

**Pull request merged**